flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/interval/SemiAntiTimeIntervalJoin.java

```diff
@@ -0,0 +1,450 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.operators.join.interval;
+
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.apache.flink.api.common.state.MapState;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.StateTtlConfig;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.typeutils.MapTypeInfo;
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.functions.co.KeyedCoProcessFunction;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.runtime.generated.JoinCondition;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+import org.apache.flink.types.RowKind;
+import org.apache.flink.util.Collector;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.table.runtime.util.StateConfigUtil.createTtlConfig;
+
+/**
+ * A CoProcessFunction to execute time interval (time-bounded) stream semi/anti-join. Two kinds of
+ * time criteria: "L.time between R.time + X and R.time + Y" or "R.time between L.time - Y and
+ * L.time - X" X and Y might be negative or positive and X <= Y.
+ */
+public abstract class SemiAntiTimeIntervalJoin
```

Comment:
It seems `SemiAntiTimeIntervalJoin` has a lot of same code with `TimeIntervalJoin`.
Could we try to use `TimeIntervalJoin`?
There is only three difference between Semi/Anti with other join type:
1. compose emit results, semi/Anti join only keep the left side.
2. For semiJoin, the join behavior is very similar with inner join, they all emit result when left side comes. But right side input of semi join would not trigger emit behavior.
3. For anti Join, the join behavior is special. AntiJoin would only emit results in two case:
     1. Left input comes, it finds out operator time of right stream has already exceeds the upper window bound of the current row, it could trigger emit result if it finds no matched row in the right side
     2. when clean timer on left side is triggered,  it could trigger emit result if it finds no matched row in the right side

Reply:
The main difference is the following:
1. different states with left and right stream.
2. different logic in `processElement1` and `processElement2`
It seems that if I use the existing strategy in `TimeIntervalJoin` to clean up states, the most methods can be reused just like you say. So maybe I can get an higher abstract class named `TimeIntervalJoin` to keep the most codes and transfer the remaining codes in `TimeIntervalJoin` to `CommonTimeIntervalJoin`. And do semi / anti join in `SemiAntiTimeIntervalJoin`. What do you think?

Reply:
Good idea

Reply:
The state structure can be unified, using Map instead List can reduce duplicate data in the state. After the state structure is unified, most process logic can be reused